stan/math/prim/scal/meta/is_vector_like.hpp

```diff
@@ -2,12 +2,21 @@
 #define STAN_MATH_PRIM_SCAL_META_IS_VECTOR_LIKE_HPP
 
 #include <stan/math/prim/scal/meta/bool_constant.hpp>
+#include <stan/math/prim/scal/meta/is_detected.hpp>
 #include <stan/math/prim/scal/meta/is_eigen.hpp>
 #include <stan/math/prim/scal/meta/is_vector.hpp>
 #include <type_traits>
 
 namespace stan {
 
+namespace internal {
+/**
+ * @brief Used to detect if object has operator[](int) defined
+ */
+template <typename T>
+using operator_bracket_t = decltype(std::declval<T>()[int{}]);
+}  // namespace internal
```

Comment:
Will this also need to account for ```operator[](size_t)```?

Reply:
Good catch! So what happens here is because `size_t` is implicitly convertible to int the above works fine and dandy. I added a test for this in `is_vector_list_test.cpp` that checks that tests pass for `size_t` and `unsigned int` `operator[]`